staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go

```diff
@@ -67,17 +73,56 @@ var (
 		},
 		[]string{"transformation_type"},
 	)
+
+	// keyIDTotal is the number of times a keyID is used
+	// e.g. apiserver_envelope_encryption_keyid_total counter
+	// apiserver_envelope_encryption_keyid_total{key_id="sha256",
+	// provider_name="providerName",transformation_type="from_storage"} 1
+	KeyIDTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "keyid_total",
+			Help:           "Number of times a keyID is used split by transformation type and provider.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"transformation_type", "observer_type", "provider_name", "key_id"},
+	)
+
+	// keyIDLastTimestamp is the last time a keyID was used
+	// e.g. apiserver_envelope_encryption_keyid_last_timestamp{key_id="sha256"} 1.674865558833728e+09
+	KeyIDLastTimestamp = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      namespace,
+			Subsystem:      subsystem,
+			Name:           "keyid_last_timestamp",
+			Help:           "The last time a keyID was used.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"key_id"},
+	)
 )
 
 var registerMetricsFunc sync.Once
 
 func RegisterMetrics() {
 	registerMetricsFunc.Do(func() {
-		legacyregistry.MustRegister(dekCacheFillPercent)
-		legacyregistry.MustRegister(dekCacheInterArrivals)
+		legacyregistry.MustRegister(DekCacheFillPercent)
+		legacyregistry.MustRegister(DekCacheInterArrivals)
+		legacyregistry.MustRegister(KeyIDTotal)
+		legacyregistry.MustRegister(KeyIDLastTimestamp)
 	})
 }
 
+// RecordKeyID records count of KeyID used for TransformFromStorage and TransformToStorage operations.
+func RecordKeyID(transformationType, observerType, providerName, keyID string) {
+	h := sha256.New()
+	h.Write([]byte(keyID))
+	keyIDHash := fmt.Sprintf("sha256:%x", h.Sum(nil))
```

Comment:
should we use `hashPool` for this: https://github.com/kubernetes/kubernetes/blob/97ab147537f6999249c241e0896245bd961c32ae/staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go#L81-L84

xref: https://github.com/kubernetes/kubernetes/pull/115350